main.py

```diff
@@ -236,104 +260,147 @@ def main(cl_arguments):
     else:
         # Look for eval checkpoints (available only if we're restoring from a run that already
         # finished), then look for training checkpoints.
-        eval_best = glob.glob(os.path.join(args.run_dir,
-                                           "model_state_eval_best.th"))
-        if len(eval_best) > 0:
-            load_model_state(
-                model,
-                eval_best[0],
-                args.cuda,
-                task_names_to_avoid_loading,
-                strict=strict)
+        best_path = get_best_checkpoint_path(args.run_dir)
+        if best_path:
+            load_model_state(model, best_path, args.cuda, task_names_to_avoid_loading,
+                             strict=strict)
         else:
-            macro_best = glob.glob(os.path.join(args.run_dir,
-                                                "model_state_main_epoch_*.best_macro.th"))
-            if len(macro_best) > 0:
-                assert_for_log(len(macro_best) == 1,
-                               "Too many best checkpoints. Something is wrong.")
-                load_model_state(
-                    model,
-                    macro_best[0],
-                    args.cuda,
-                    task_names_to_avoid_loading,
-                    strict=strict)
-            else:
-                assert_for_log(
-                    args.allow_untrained_encoder_parameters,
-                    "No best checkpoint found to evaluate.")
-                log.warning("Evaluating untrained encoder parameters!")
+            assert_for_log(args.allow_untrained_encoder_parameters,
+                           "No best checkpoint found to evaluate.")
+            log.warning("Evaluating untrained encoder parameters!")
+            if args.transfer_paradigm == "finetune":
+                # Save model so we have a checkpoint to go back to
+                # after each task-specific finetune.
+                model_state = model.state_dict()
+                model_path = os.path.join(args.run_dir, "model_state_untrained_prefinetune.th")
+                torch.save(model_state, model_path)
 
     # Train just the task-specific components for eval tasks.
     if args.do_target_task_training:
-        # might be empty if no elmo. scalar_mix_0 should always be pretrain scalars
-        elmo_scalars = [(n, p) for n, p in model.named_parameters() if
-                        "scalar_mix" in n and "scalar_mix_0" not in n]
-        # fails when sep_embs_for_skip is 0 and elmo_scalars has nonzero length
-        assert_for_log(not elmo_scalars or args.sep_embs_for_skip,
-                       "Error: ELMo scalars loaded and will be updated in do_target_task_training but "
-                       "they should not be updated! Check sep_embs_for_skip flag or make an issue.")
+        if args.transfer_paradigm == "frozen":
+            # might be empty if elmo = 0. scalar_mix_0 should always be pretrain scalars
```

Comment:
Rephrase to make it clearer that it should be empty if sep_embs_for_skip is 0.

Reply:
It can be non-empty even if `sep_embs_for_skip = 0` since there are still target-task specific mixing weights.

Reply:
I'm confused. If `sep_embs_for_skip = 0`, isn't it possible to have elmo scalars to train (task-specific layer mixing weights)?

Reply:
Oh, there are no other ELMo scalars if no skip connection b/c there are no task-specific weights in that case?

Reply:
Er, I probably just misunderstood this. Nvm.